src/Scheduler/GenericScheduler.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console\Scheduler;
+
+use DateTime;
+use Tempest\Console\ConsoleCommand;
+use Tempest\Console\ConsoleOutput;
+
+final class GenericScheduler implements Scheduler
+{
+    public function __construct(
+        private SchedulerConfig $config,
+        private ConsoleOutput $output,
+        private ScheduledCommandsResolver $resolver,
+    ) {
+    }
+
+    public function run(?DateTime $date = null): void
+    {
+        $date ??= new DateTime();
+
+        $commands = $this->resolver->resolve($date);
+
+        foreach ($commands as $command) {
+            $this->execute($command);
+        }
+    }
+
+    private function execute(ConsoleCommand $scheduledCommand): void
+    {
+        $name = $scheduledCommand->getName();
+        $this->output->writeln("Running command: $name");
+
+        $command = $this->compileCommand($scheduledCommand);
+
+        exec($command);
+
+        $this->output->writeln("Command finished: $name");
```

Comment:
I don't think it's finished at this point, since the command is run in the background. Running them as background jobs actually opens a huge can of worms, because now we need to keep track of running, succeeded, and failed commands as well.

That in turn leads to this feature becoming a queue manager instead of a simple scheduler. I think that's out of scope. (I do plan on writing a queue manager, but not within this package)

I think we should always run scheduled commands in the same process. But I might overlook some problems with that approach. I'd appreciate your input

Reply:
It should say "dispatched" instead, but we don't really want this output here either way.

> I don't think it's finished at this point, since the command is run in the background. Running them as background jobs actually opens a huge can of worms, because now we need to keep track of running, succeeded, and failed commands as well.

I may be missing something here, but from the point of scheduler, we are not really interested in whether or not they succeed - we just want stuff to run, or really - be dispatched at specific time, if we were to execute them in sync, we'd have to deal with delays and stuff not running on time.

I think having them run in background should be the default option, defined in config.

I think that eventually, what would be cool is if we could make the Scheduler dispatch n-tasks to execute commands in parallel that'd report their status via events - this way we could have it work kinda both ways

Reply:
> if we were to execute them in sync, we'd have to deal with delays and stuff not running on time.

That's a good point 🤔

I'm just a little scared about things being dispatched to the background, and not having any way of interacting with them anymore. What if a process hangs indefinitely? We definitely need some good logging then. 

Actually. Writing this down. It would be pretty powerful if the scheduler and queue manager shared the same infrastructure. I always found it a bit confusing with Laravel that there is a basic scheduler, and then there's horizon as something separate.

I'm not saying we should write that queue management part right now, but it might be an interesting area to explore afterwards.

Ok, running in the background makes most sense. Thanks for sharing your thought!